File: plugin-sdk-ts/src/http.ts
import { callHostJson } from './host';
import type { HttpRequest, HttpResponse } from './types';

/**
 * Make an HTTP request (must be in allowed domains).
 * Requires: http permission + domain allowlist.
 */
export function request(input: HttpRequest): HttpResponse {
  return callHostJson<HttpResponse>(
    'cognia_http_request',
    JSON.stringify({
      method: input.method ?? 'GET',
      url: input.url,
      headers: input.headers ?? {},
      body: input.body,
      contentType: input.contentType,
      timeoutMs: input.timeoutMs,
    }),
  );
}

/**
 * Compatibility helper for a simple HTTP GET request.
 * For new plugins, prefer `request()` to use the normalized request contract.
 */
export function get(url: string): HttpResponse {
  return request({ method: 'GET', url });
}

/**
 * Compatibility helper for a simple HTTP POST request.
 * For new plugins, prefer `request()` to use the normalized request contract.
 */
export function post(
  url: string,
  body: string,
  contentType?: string,
): HttpResponse {
  return request({
    method: 'POST',
    url,
    body,
    contentType: contentType ?? 'application/json',
  });
}

File: plugin-sdk/src/http.rs
use crate::host;
use crate::types::{HttpRequest, HttpResponse};
use extism_pdk::*;

/// Make an HTTP request (must be in allowed domains).
pub fn request(input: &HttpRequest) -> Result<HttpResponse, Error> {
    let payload = serde_json::json!({
        "method": input.method.clone().unwrap_or_else(|| "GET".to_string()),
        "url": input.url,
        "headers": input.headers.clone().unwrap_or_default(),
        "body": input.body,
        "contentType": input.content_type,
        "timeoutMs": input.timeout_ms,
    })
    .to_string();
    let result = unsafe { host::cognia_http_request(payload)? };
    Ok(serde_json::from_str(&result)?)
}

/// Make an HTTP GET request to a URL (must be in allowed domains).
pub fn get(url: &str) -> Result<HttpResponse, Error> {
    request(&HttpRequest {
        url: url.to_string(),
        method: Some("GET".to_string()),
        headers: None,
        body: None,
        content_type: None,
        timeout_ms: None,
    })
}

/// Make an HTTP POST request (must be in allowed domains).
pub fn post(url: &str, body: &str, content_type: Option<&str>) -> Result<HttpResponse, Error> {
    request(&HttpRequest {
        url: url.to_string(),
        method: Some("POST".to_string()),
        headers: None,
        body: Some(body.to_string()),
        content_type: Some(content_type.unwrap_or("application/json").to_string()),
        timeout_ms: None,
    })
}
